### ISSUE TITLE
How to set the line spacing and paragraph spacing to 1zw (zw=full-width character width)

### ISSUE THREAD

**Issue by Yoxem** [sile-typesetter/sile#521]
I want to modified jplain.lua to make a chinese plain template.
Because in Requirements for Chinese Text Layour (w3c), the spacing between lines and that between paragraphs should be the same (cf. In principle, there is no extra space between paragraphs in Chinese books.) I want to try to implement it in the zplain.lua, but I failed. 

The following are some lines in zplain.lua:

```Lua
-- zw = the width of a full-width character
SILE.registerUnit("zw", { relative = true, definition = function (v)
    return v * SILE.shaper:measureChar("永").width
end})

-- setting the indent

SILE.languageSupport.loadLanguage("zh")
SILE.settings.set("document.parindent",SILE.nodefactory.newGlue("2zw"))

-- modified the linespacing
 toAdd = SILE.length.new({ length = SILE.toPoints("2zw")})
 SILE.nodefactory.newVglue({ height = toAdd })

return zplain
```

How can I fix it and make it work after compiling and installing (i.e. how can make it be readable by the path)

[zplain.lua.zip](https://github.com/simoncozens/sile/files/1781392/zplain.lua.zip)
[zh.lua.zip (workaround, borrowed from ja.lua)](https://github.com/simoncozens/sile/files/1781439/zh.lua.zip)
[zplain-test.pdf](https://github.com/simoncozens/sile/files/1781447/zplain-test.pdf)
[zplain-test.sil.zip](https://github.com/simoncozens/sile/files/1781450/zplain-test.sil.zip)

**Comment by Yoxem** [comment]
I've found that SILE.settings.set("document.parskip" , SILE.nodefactory.newVglue("[width]")) should work, but it didn't work. see: https://github.com/Yoxem/sile/blob/master/classes/zplain.lua (Line 35)

**Comment by Yoxem** [comment]
I add this line to zplain.lua. eg:
  
-- modified the linespacing and parskip
SILE.settings.set("document.parskip" , SILE.nodefactory.newVglue(SILE.toPoints("0zw").."pt plus 1pt")) 
SILE.settings.set("linespacing.method","fixed")
SILE.settings.set("linespacing.fixed.baselinedistance",SILE.length.new({length = SILE.toPoints("3zw")}))
-- modified the linespacing and parskip
SILE.settings.set("document.parskip" , SILE.nodefactory.newVglue(SILE.toPoints("0zw").."pt")) -- It seems that it doesn't work
SILE.settings.set("linespacing.method","fixed")
SILE.settings.set("linespacing.fixed.baselinedistance",SILE.length.new({length = SILE.toPoints("3zw")}))
print(SILE.settings.get("document.parskip"))
return zplain

the output result in the command is like:
This is SILE 0.9.5-unreleased
<test.sil>
VG<0pt>

! Overfull frame: 27.3900625 extra space required but 0 shrink available
[1] [2]

but it doesn't work.

However, when I added "\set[parameter=document.parskip,value=-0zw]" to the test .sil file,
the parskip setting did work.

Is this a bug of SILE?

**Comment by simoncozens** [comment]
I think I understand what is happening. The `hanmenkyoshi` package, which is responsible for setting up the CJK grid (版面), also sets up line spacing (because it computes the size of the frame based on the number of lines plus the line gap). It sets both the paragraph skip and the line skip. Because this happens at class init time (in the `init` method of your class - which is where it should be) this overrides the settings you make when your class is loaded.

However, the grid which `hanmenkyoshi` sets up is perhaps not flexible enough. It takes two parameters, grid size and line gap. This sets the line skip correctly but it assumes that the paragraph skip should be grid size + line gap. I don't know where I got this from. It might be wrong.

One easy way to work around it is to set `document.parskip` inside the `init` method and after the call to `declareHanmenFrame`:

```lua
  self:declareHanmenFrame( "content", {
    left = "8.3%pw", top = "11.6%ph",
    gridsize = "1zw", linegap = "1zw", linelength = 50,
    linecount = 30,
    tate = self.options.layout() == "tate"
  })
  SILE.settings.set("document.parskip", SILE.nodefactory.newVglue("1zw"))
```

I think the real fix is for me to go back to the W3C Requirements and check if we are declaring the grid correctly. in `hanmenkyoshi`.

**Comment by simoncozens** [comment]
By the way, would you mind it if we add `zh.lua` and `zplain.lua` to the repository?

**Comment by Yoxem** [comment]
Simon Cozens <notifications@github.com> 於 2018年3月20日 週二 06:30 寫道：

**Comment by Yoxem** [comment]
Sure, but the code may be needed to modify.

Simon Cozens <notifications@github.com> 於 2018年3月20日 週二 06:30 寫道：

**Comment by simoncozens** [comment]
OK, after re-reading jlreq, I understand what I was trying to do with the parskip, but I think I got it wrong. Paragraphs should simply start on the next available new line on the grid. But of course they were going to do that anyway, so there should not be any additional skip. I've set parskip to 0 in hanmenkyoshi.

**Comment by Omikhleia** [comment]
It is very unclear if anything else was expected after the last comment, or if this ticket is actually closed...

**Comment by ctrlcctrlv** [comment]
It should be closed. For CJK a parskip of 1zw is normative, including for Japanese, the CJK language I write.

![Screenshot from 2022-11-08 18-06-02](https://user-images.githubusercontent.com/838783/200695251-defb3b12-2899-444b-8109-051914c99741.png)

**Comment by alerque** [comment]
Fix was in v0.9.0.
